### DISCUSSION TITLE
Skip table creation for initial migration? | Adding flask-migrate to existing project

### DISCUSSION THREAD

**Discussion by thilojaeggi** [miguelgrinberg/Flask-Migrate#516] [Q&A]
So, my current project is already out there in a docker image without using Flask-Migrate.

What I've done now is:

1. Created an empty database
2. Ran flask db init
3. Ran flask db migrate
4. Made the changes in the models
5. Ran flask db migrate again

Now I have the first migrate which creates all the tables, but the users already have these tables so this fails, subsequently the following migration will fail as well/never get ran.

What I've now done is this workaround, but I wanted to ask if there's a better approach:

```
"""empty message

Revision ID: 3d1965dfdcd5
Revises: 
Create Date: 2023-07-01 11:23:57.334012

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '3d1965dfdcd5'
down_revision = None
branch_labels = None
depends_on = None

inspector = sa.inspect(op.get_bind())


def upgrade():
    table_names = inspector.get_table_names()
    if 'package' in table_names and \
       'package_version' in table_names and \
       'installer' in table_names and \
       'installer_switch' in table_names:
        print('All Tables already exist, skipping creation')
        return
```
this works perfectly fine, because of this I also changed the docker entry point to be this:
```
#!/bin/sh
flask db upgrade
exec gunicorn -b :8080 --workers 4 "app:create_app()"
```

**Comment by miguelgrinberg** [comment]
You don't need to modify your migration scripts for this. Just ask your users to run this command to register their databases as up to date:

    flask db stamp head

**Reply by thilojaeggi** [reply]
But if I already have the second migration in there for changing the column name, then it will register that it has already been renamed and therefore not do it, no?

**Reply by miguelgrinberg** [reply]
Okay, sorry, the correct thing here is to use

    flask db stamp <id-of-first-migration>
    flask db upgrade

The stamp command will make their databases as in sync with the first migration, then the upgrade will apply the second migration.

**Reply by thilojaeggi** [reply]
Okey, so basically I'll just add a check if the database exists (somehow, don't know yet) and then if it does, run flask db upgrade?

Or when it detects no db the flask db stamp will automatically fail and go on to the first migration?

I'd also have to change/remove this from the next release after this I would assume so this is basically requires users when upgrading from the current version to go through this version and not be allowed to skip it.

**Reply by miguelgrinberg** [reply]
You keep trying to build this logic in the migrations, but migrations are not supposed to be used in that way, at least I wouldn't recommend that you do. Migrations just describe changes to the database. That's it.

What you need to do is tell all the people who have databases that predate the migrations to run the `flask db stamp` command to sync their databases, then run the upgrade command regularly to incorporate database changes from then on.
